[WIP] Add support for GitHub actions

jaimergp opened the pull request:
This PR adds preconfigured support for GitHub Actions, following conversation in #90.

# To do

- [X] Parameterize existing implementation in `openkinome/kinoml`
- [x] Add support for Windows
- [x] Add support for non-conda builds
- [x] Apparently [Miniconda is preinstalled in all environments](https://help.github.com/en/actions/automating-your-workflow-with-github-actions/software-installed-on-github-hosted-runners), the location being available at `$CONDA`. Test this is true and functional.
- [x] Variables in GH Actions use some sort of Jinja-syntax (`${{ variable }}`) that interferes with cookiecutter's Jinja processing. I think we just need to escape this on the unprocessed YAML to make the tests pass.

dgasmith commented on {{cookiecutter.repo_name}}/.github/workflows/CI.yaml:
```diff
@@ -0,0 +1,45 @@
+        # stop the build if there are Python syntax errors or undefined names
+        flake8 . --count --select=E9,F63,F7,F82 --show-source --statistics
+        # exit-zero treats all errors as warnings. The GitHub editor is 127 chars wide
+        flake8 . --count --exit-zero --max-complexity=10 --max-line-length=127 --statistics
```
Can we turn the flake items off as we do not recommend them at the moment.

dgasmith commented on {{cookiecutter.repo_name}}/.github/workflows/CI.yaml:
```diff
@@ -0,0 +1,45 @@
+      max-parallel: 4
+      matrix:
+        os: [macOS-latest, ubuntu-latest]
+        python-version: [3.6, 3.7, 3.8]
```
As noted we might want to turn off 3.8 for now until c-f finishes their builds.

dgasmith commented on {{cookiecutter.repo_name}}/.github/workflows/CI.yaml:
```diff
@@ -0,0 +1,45 @@
+        python devtools/scripts/create_conda_env.py -n=test -p={{ '${{ matrix.python-version }}' }} devtools/conda-envs/test_env.yaml
+    - name: Install package
+      run: |
+        eval "$(${CONDA}/conda shell.bash hook)" && conda activate test
```
Too bad we cannot outrun this. Can we do something like:

```bash
##vso[task.setvariable variable=CONDA_ENV]"$(${CONDA}/conda shell.bash hook) && conda activate test"
eval $CONDA_ENV
```

so that it will be easier to add additions steps?

jaimergp replied:
We would still need to `eval $CONDA_ENV` and `conda activate test` in all steps, right? I'd rather be explicit about and not use this Azure magics (where can I find docs about them, btw?) if possible.

jaimergp replied:
Found the equivalent docs for GHA: https://help.github.com/en/actions/automating-your-workflow-with-github-actions/development-tools-for-github-actions#set-an-environment-variable-set-env

I am using that to define `CONDA_EXE` across different platforms. We still need to initialize/eval the shell and activate the environments, though.

dgasmith commented on README.md:
```diff
@@ -16,7 +16,7 @@ remove deployment platforms, or test with a different suite.
 * Python-centric skeletal structure with initial module files
 * Pre-configured `setup.py` for installation and packaging
-* Pre-configured Window, Linux, and OSX continuous integration on AppVeyor and Travis-CI
+* Pre-configured Window, Linux, and OSX continuous integration on AppVeyor+Travis-CI and/or GitHub Actions.
```
I am up for removing AppVeyor and Travis entirely. @janash @Lnaden any comments?

Lnaden replied:
I am all for dropping Travis and VERY much for dropping AppVeyor. Go for it!

jaimergp replied:
Shall I remove them as part of this PR or should this be in a separate PR (also merged to `dev`) for extra clarity?

jaimergp commented:
I have added Windows support by simply forcing `bash` to be the default shell in all steps. This is technically supported by GHA, but I'd like to test it with a dummy package first. I think I can use `openkinome/kinoml` for that in the meantime.

jaimergp commented:
That seemed to work. Look here for a similar setup: https://github.com/openkinome/kinoml/pull/5/checks

jaimergp commented:
By the way, do we want to update the CI on this repo too so it uses GH Actions?

dgasmith commented:
This looks really great, let me try this out in a few of our packages and see how it goes.

I would vote to changing the repo's CI in another PR.

dgasmith commented:
This is my pass at the YAML: https://github.com/MolSSI/QCElemental/pull/172/files

dgasmith commented on {{cookiecutter.repo_name}}/.github/workflows/CI.yaml:
```diff
@@ -0,0 +1,74 @@
+{% if cookiecutter.dependency_source != 'Dependencies from pip only (no conda)' %}
+        eval "$(${CONDA_EXE} shell.bash hook)" && conda activate test
+{% endif %}
+        codecov
```
BTW this doesn't work without a token. I recommend using their image

```
    - name: CodeCov  
       uses: codecov/codecov-action@v1
       with:
         token: ${{ secrets.CODECOV_TOKEN }}
         file: ./coverage.xml
         flags: unittests
         yml: ./.codecov.yml 
```

Add ` --cov-report=xml` to the pytest line as well.

jaimergp replied:
I just noticed and was about to ask! Thanks!

dgasmith commented on {{cookiecutter.repo_name}}/.github/workflows/CI.yaml:
```diff
@@ -68,7 +68,7 @@ jobs:
       uses: codecov/codecov-action@v1
       with:
-        token: ${{ secrets.CODECOV_TOKEN }}
+        token: {{ '${{ secrets.CODECOV_TOKEN }}' }}
```
I don't think the braces are needed here? This isn't needed on my yaml.

jaimergp replied:
We need to escape them so `cookiecutter`'s Jinja processing doesn't parse it. It's also needed on `runs-on`, for example.

dgasmith replied:
Ah right, its rendered...

dgasmith commented on {{cookiecutter.repo_name}}/.github/workflows/CI.yaml:
```diff
@@ -0,0 +1,74 @@
+      shell: bash
+      if: startsWith(matrix.os, 'windows')
+      run: echo "::set-env name=CONDA_EXE::$CONDA/Scripts/conda.exe"
+{% endif %}
```
Please do look through this YAML, a lot of this simply isn't needed: https://github.com/MolSSI/QCElemental/blob/827e3e6faf9ba08f327bafff87e893b8f8d559fa/.github/workflows/CI.yml

jaimergp replied:
So `conda` is in `PATH`? Cool!

jaimergp replied:
Looks like this is not applicable to Windows! Shall I revert the changes?

dgasmith replied:
Maybe add it to the path for Windows only?

jaimergp replied:
I followed [this discussion](https://github.com/actions/starter-workflows/issues/44#issuecomment-554649325) and there seems to be some problems with path modification. I'll try to use `condabin` instead, and see if that works better.

jaimergp replied:
Actually, [looks like](https://github.com/openkinome/kinoml/pull/7/checks?check_run_id=328385167) `conda` is only in `PATH` in Ubuntu. I keep thinking that the `CONDA_EXE` approach is the cleanest although it requires one more setup step.

We could use a single step too:

```
- name: Set-up conda
        run:   |
          if [ "${{ matrix.os }}" != "windows-latest" ]; then
            echo '::set-env name=CONDA_EXE::${CONDA}/bin/conda'
          else
            echo '::set-env name=CONDA_EXE::C:\Miniconda\condabin\conda.bat'
          fi
        shell: bash
```

dgasmith replied:
Do you think we could add a conda function in this manner? If not I would recommend something simple so that further lines can do `eval '$CONDA_INIT'` or similar.

jaimergp replied:
Best thing would be modifying the `bash_profile`/`bashrc` to autostart conda for us (but we should still run `conda activate test` once the environment has been created, although we could check for `test` existence and activate that instead?), but I think that default `bash` interpreters in GitHub are [configured to ignore profile files](https://help.github.com/en/actions/automating-your-workflow-with-github-actions/workflow-syntax-for-github-actions#using-a-specific-shell), so we would need to define a custom `shell` in each step. I am not aware of any option to override the default `shell` for all steps - if that's possible, it'd be the best way. Anyway, shall I try with the `bashrc`+`custom bash shell` approach?

jaimergp replied:
Ok, I've tried this route but it's not pretty at all because we would need to _always_ specify `shell: bash --noprofile --rcfile conda_init.sh -i -e -o pipefail {0}`. I find that very confusing and hard to reason about. 

Instead, I have opted to create [this auxiliary script](https://github.com/openkinome/kinoml/blob/329f764b780e9bcfa228a972625ab3ee7e840729/devtools/scripts/initialize_conda.sh) and call it at the beginning of each step:

```
    - name: Install package
      shell: bash
      run: |
        . devtools/scripts/initialize_conda.sh
        python -m pip install . --no-deps
        conda list
```

That way the intention is clear and the initialization code is easier to maintain. Would you prefer this approach?

jaimergp replied:
I went ahead and added the script in the last commit.

dgasmith commented on {{cookiecutter.repo_name}}/.github/workflows/CI.yaml:
```diff
@@ -0,0 +1,74 @@
+      run: |
+        eval "$(${CONDA_EXE} shell.bash hook)" && conda activate
+        conda info
+        python devtools/scripts/create_conda_env.py -n=test -p={{ '${{ matrix.python-version }}' }} devtools/conda-envs/test_env.yaml
```
I like going through ENV variables for these, makes it much easier.

jaimergp replied:
Nice! Applied as suggested.

dgasmith commented:
On the codecov line add `--color yes`. Unsure why color isn't shown without this.

jaimergp commented:
Oh, due to a [WSL bug](https://github.com/microsoft/WSL/issues/4245), my commits are dated as 4 days old. Next ones should be ok, though. Sorry for the confusion!

dgasmith commented:
Hmm, we do have a hang up. Because GHA Codecov requires secrets it doesn't let us get a codecov update on pull requests.

jaimergp commented:
How is this solved with other CIs?

dgasmith commented:
https://docs.codecov.io/docs/about-the-codecov-bash-uploader#section-upload-token

I assume codecov looks at the domain of the upload and accepts the packet. Azure DevOps apparently works, unsure why GHA doesn't or if this is being considered.

jaimergp commented:
Also, I'd like to reconsider the use of `CONDA_EXE` env var. In Windows, `conda` is apparently not in `PATH`. See [here](https://github.com/openkinome/kinoml/pull/7/checks?check_run_id=328172280).

jaimergp commented:
About codecov tokens:

* This [comment](https://github.com/codecov/codecov-python/issues/209#issuecomment-533898306) proposes two workarounds for `codecov-python`. There's also a [PR](https://github.com/codecov/codecov-python/pull/214).
* `codecov-bash` has [support for GHA](https://github.com/codecov/codecov-bash#ci-providers) already. Not sure if it's tokenless, though.

dgasmith commented:
Yea, looks like we need to wait for this to merge. The codecov-bash says it needs tokens for both public and private instances.

dgasmith commented on {{cookiecutter.repo_name}}/devtools/scripts/initialize_conda.sh:
```diff
@@ -0,0 +1,16 @@
+# $CI_OS is $matrix.os, as exported in GHA *.yaml
```
This should be in the `gha` directory similar to `travis` I think.

jaimergp replied:
Ok! I named it `github-actions` following the scheme in `travis-ci`.

dgasmith commented on {{cookiecutter.repo_name}}/devtools/scripts/initialize_conda.sh:
```diff
@@ -0,0 +1,16 @@
+        eval "$(conda shell.bash hook)";;
+esac
+
+if [[ -d ${CONDA}/envs/test ]] || [[ -d ${HOME}/.conda/envs/test ]]; then
```
What if this took in args, so you could `devtools/scripts/conda.sh active test`?

jaimergp replied:
Good idea. I renamed the script so it's named `activate_conda_env.sh`, which will accept `$1`. If none is provided, it will run `conda activate`, but if you say `activate_conda_env.sh test` it will run `conda activate test`. Does that make sense?

dgasmith replied:
But why do that? If you allow the bash script take in arbitrary numbers of arguments then this could be used for `conda list` or other operations as well.

jaimergp replied:
Once you have called the script, you can execute `conda` normally:

```
. devtools/github-actions/activate_conda_env.sh
# base environment is now active
conda create -n test python=3.8 my_package
conda list
```

Would you like the activation step to be separate and explicit from the initialization? As in:

```
. devtools/github-actions/initialize_conda.sh
conda activate
```

dgasmith replied:
Bleh, I get that they do not particular like to keep state. But at the same time they are making it really hard to make a few operations easy. I vote a bit for `. devtools/github-actions/initialize_conda.sh` as it seems clear.

jaimergp replied:
One of the things I'd like to see implemented is defining custom shells per job, so you can do:

```
jobs:
  test:
    runs-on: {{ '${{ matrix.os }}' }}
    strategy:
      matrix:
        os: [macOS-latest, ubuntu-latest, windows-latest]
        python-version: [3.6, 3.7]
        bash-with-conda: ["bash --rcfile ~/devtools/github-actions/initialize_conda.sh"] 

[...]

    - name: Install testing dependencies
      shell: ${{ matrix.bash-with-conda }}
      run: |
        python -m pip install -U pytest pytest-cov codecov
```

This syntax is correct, but `shell` is not aware of the `${{ }}` expansions (yet?). For now we will need to resort to this script. I understand that we will have `initialize_conda.sh` only handling the pre-activation routines and then being explicit about further steps like `conda activate test`?

dgasmith replied:
Does `echo "::add-path::/path/to/dir"` help here?

jaimergp replied:
We _could_ add `$CONDA/condabin` to path to have the `conda` executables in `PATH`, but we still need to `eval` the bash hook and activate the environment anyway, so I don't see the need to pollute the path.

dgasmith replied:
If we added it to the path `conda init bash` should work across the board? Its mostly pedantic I know, but it looks much more similar to the operations you would complete on a local computer.

jaimergp replied:
Apparently the consensus is to use the bash thingy now: https://github.com/conda/conda/issues/7980#issuecomment-492784093. `conda init bash` modifies your `.bashrc` so this hook is automatically executed in every session. However, GHA's `bash` does not load profiles, and we would have to specify our own custom flags in the `shell` option anyway... I wish this could be simpler, but so far this (script) is the best option I've found.

SimonBoothroyd replied:
Is there a reason not to just use something like https://github.com/s-weigand/setup-conda to simplify this?

jaimergp replied:
It adds `$CONDA/bin` to `$PATH`, which is a bit too invasive depending on how you configure your `base` environment. This approach also requires the old `source activate <env>` mechanism, which is supposed to be deprecated (but still working). Anyway, it could stop working anytime.

The recommended way is using `eval` on the Bash hooks (as we are doing), per process. One cleaner workaround would be adding `$CONDA/condabin` (only contains `conda`) to `$PATH`, which we haven't tested, but again this would require `source activate`.

The best solution would be a mechanism in GHA to have each interpreter run the same command upon startup. If they didn't ignore `.bashrc` and friends, that is.

SimonBoothroyd replied:
I was more curious about something *like* that - i.e come up with a new one and release on Marketplace that fits the spec, or just modify the existing one, rather than advocating using *that* exact one.

jaimergp replied:
Oh, I see what you mean. The problem with releasing this as a GH Action is that we still need to execute it for each step because state is not kept across steps, and I don't think actions support that kind of behavior. I wonder if an Action can define a custom shell though; like "subclassing" their Bash shell with our own action and the using that... That'd be the cleanest, IMO.

SimonBoothroyd replied:
> still need to execute it for each step because state is not kept across steps
Ah perhaps this is a misunderstanding on my part - I though that state set by an action is then kept across all steps within the job, such that the environment should stay active throughout, but I could be wrong about this!

jaimergp replied:
Unfortunately each step is a separate Bash/shell process, so that kind of state is lost. Only thing you can keep across states is `$PATH` and custom environment variables, which is enough for `conda` to work with `source activate` but not the recommended approach (for some reason).

jaimergp commented:
> Yea, looks like we need to wait for this to merge. The codecov-bash says it needs tokens for both public and private instances.

Finally found the [thread](https://community.codecov.io/t/whitelist-github-action-servers-to-upload-without-a-token/491) where this is actually being discussed. They are working on it, so hopefully it won't take long.

![image](https://user-images.githubusercontent.com/2559438/70102969-c261e880-1607-11ea-87fa-dffc8f119f6f.png)

dgasmith commented:
What do you think about sitting on this for a bit until we 1) have a versioning plan for the cookiecutter and 2) let items like codecov catchup?

jaimergp commented:
Yes, sure, no problems. I'll keep an eye for updates on CodeCov tokenless integration. We will probably be able to include Python 3.8 by then too.

jaimergp commented:
In the meantime, should I do anything else about this? https://github.com/MolSSI/cookiecutter-cms/pull/95#discussion_r353508399

dgasmith commented:
@dotsdl your feedback may be useful here!

jaimergp commented:
Re Codecov, here's an update:

![image](https://user-images.githubusercontent.com/2559438/70394402-28c67c80-19f5-11ea-819d-94635500bd39.png)

So, it will take a bit longer than expected. Do we want to wait that long or maybe we can compromise on a short term solution? E.g. merging and providing both Travis/AppVeyor and GHA? (I don't mind waiting - I can always get the template from this PR if a project requires a quicker integration).

dgasmith commented:
What if we merge this into a "dev" branch?

I don't really understand their response. There is an open PR that implements this exact functionality as it currently stands. Unless that PR uses "alpha" features, have seen anywhere GHA labels different features at different status levels?

jaimergp commented:
In a previous message in that thread: 

> Tokenless uploads works by fetching and confirming the build via the API. I’m not sure if this is possible with GitHub Actions, but I’ve passed the request along to our engineering team.

Until GHA provides an API to check the builds, tokenless are not possible. This is a change on codecov server-side. The opened PRs only implement the with-token methods in the clients, I think.

jaimergp commented:
Oh, and I'm fine with merging into `dev`. Would that be the standard release process from now on? A staging branch with latest changes? I feel that that kind of change would need to be documented first. If this is going to be a one-time thing, maybe it's enough to leave this PR open.

dgasmith commented:
Related to #97.

jaimergp commented:
GH Actions API is now beta since [Jan 28th](https://developer.github.com/changes/2020-01-28-actions-api/), which means it's tokenless codecov is coming!

![image](https://user-images.githubusercontent.com/2559438/73745657-d1722d80-4753-11ea-866e-45f104de28e3.png)

dgasmith commented:
Cool :)

hannahbrucemacdonald commented:
From @jaimergp, to @jaimergp 
Configuring a workflow - GitHub Help https://help.github.com/en/actions/configuring-and-managing-workflows/configuring-a-workflow#adding-a-workflow-status-badge-to-your-repository

- [ ] Update the homepage badge to link to GHactions, rather than travis

jaimergp commented:
That last commit should address that issue, @hannahbrucemacdonald. Thanks for the report!

dgasmith commented:
@jaimergp Any update on GHA and a place where we could track this?

jaimergp commented:
I check [this thread](https://community.codecov.io/t/whitelist-github-action-servers-to-upload-without-a-token/491/25) every Monday, but there's no need for that now because apparently it's working now! I will double check later today or tomorrow, but that means that `codecov` is no longer a blocking issue.

Do we want to add some optional linting while we are at it?

dgasmith commented:
I think we should be pretty conservative in this PR since its such a big change. We might also need to merge this into a dev branch to make sure all the kinks are ironed out.

@janash Heads up that this is now valid and we will likely proceed shifting this over soon.

dotsdl commented:
Shall we try @dgasmith's suggestion that this gets merged into a `dev` branch so that we can see how it behaves properly? Is there anyone with merge ability that can make this happen?

jchodera commented:
@lnaden @janash @bennybp are still active maintainers, as far as I know.

Lnaden commented:
I've read through all of this, what are the things we are waiting for, @jaimergp? By my thinking, we just need to change the target branch this is being merged in to? Or is there something else I am missing?

jaimergp commented:
We have been using GHA for a while now in several repositories, and have accumulated some knowledge on how to use it. I have some fixes to add here as well (macOS permissions, for example).

I'd say I can add the fixes here, and then we merge to dev. Additional features could be targeted in separate PRs. What do you think?

janash commented:
Considering that we are using CookieCutter now for MolSSI Education, I would ask that the GHA change be put on a branch or that GHA is listed as an option in CookieCutter set up (travis-ci or GHA)

jchodera commented:
I like the idea of being able to select either travis-ci or GHA until we can make sure that the GHA works robustly for a variety of people. At some point, we can just make GHA the default once folks are happy with it (and hopefully the MolSSI Education program can switch over at that point too).

Lnaden commented:
@jaimergp Is it possible for you to make your changes here that you mentioned, then once you're done, I'll come back in and make it an option to either use travis or GHA as part of the cookiecutter? That way we can get this in without having to worry about a separate branch, and we can mark the GHA experimental while we work out kinks.

jaimergp commented:
Added to my list. I'll add them some time this week!

dgasmith commented:
Big fan of:

```
    - name: Setup Conda Environment
      uses: goanpeca/setup-miniconda@v1
      with:
         activate-environment: test
         environment-file: devtools/conda-envs/test_env.yaml
         auto-activate-base: false
```

```
    - name: Environment Information
      shell: bash -l {0}
      run: |
        conda info
        conda list
```

Seems to work very well these days.

jaimergp commented:
@Lnaden, I have added the last changes I was talking about. Namely:

* Rely on a GitHub Action to deal with the Miniconda stuff, as recommended by @dgasmith.
* Added more informative job names
* Add a "good enough" general case for the push/pr branch configuration. Only commits on master and PRs against master are tested. This removes the duplicated jobs we saw in early attempts.
* Ability to choose the CI provider and update the tests accordingly

Todo:

* [ ] I don't know how to deal with the tests for GH Actions. What should we try to cover?
* [ ] Documentation must be updated to reflect these changes.

Lnaden commented:
Awesome. I think this has come together fantastically. There is just one small change I would like to request: I am thinking we should make the travis (or +appveyor) default first until we fully debug GHA. Does that work for you?

Things I intend to do after we merge:
* Add a GHA CI test
* Bump version to 3.7 and 3.8
* Reword some of the CI options so they are more clear on their supported OS's
* Update docs.

jaimergp commented:
That's an excellent point, yes! This shouldn't be the default just yet. I'll add the changes now.

Lnaden commented:
Great. Once those are in, I think we can merge this. I'll keep the 1.3 until we get GHA stable, then move to 2.0

jaimergp commented:
Hm, I am thinking that maybe we should label GH Actions as experimental in this early stage? Like this:

```json
    "continuous_integration_provider": [
        "Travis",
        "Travis+AppVeyor",
        "GitHub Actions (experimental)"
    ],
```

What do you think?

Lnaden commented:
I think thats a great idea. Makes it very clear this is a work in progress.

jaimergp commented:
Ok, let me add that before you merge.

Lnaden commented:
I'm getting one more pair of eyes to look over this before I hit merge just in case I missed something obvious, but I'm investigating the nested/chained GHA in the meantime.
